File: frontend/src/services/api/disasterRecovery.ts
/**
 * Disaster Recovery API Service
 * API client for DR management, testing, and monitoring
 */

import { apiClient } from './client';

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

export interface DRStatus {
  overall_status: string;
  health_score: number;
  components: Record<string, DRComponentStatus>;
  recent_events: DREvent[];
  recent_tests: DRTestResult[];
  last_updated: string;
}

export interface DRComponentStatus {
  status: string;
  rto_minutes: number;
  rpo_minutes: number;
  priority: number;
  automated_recovery: boolean;
  last_test_time: string | null;
  last_backup_time: string | null;
  dependencies: string[];
}

export interface DRComponent {
  component: string;
  status: string;
  rto_minutes: number;
  rpo_minutes: number;
  priority: number;
  automated_recovery: boolean;
  last_test_time: string | null;
  last_backup_time: string | null;
  dependencies: string[];
}

export interface DREvent {
  event_id: string;
  timestamp: string;
  event_type: string;
  severity: string;
  component: string;
  description: string;
  status: string;
  recovery_actions: string[];
}

export interface DRTestResult {
  test_id: string;
  test_type: string;
  component: string;
  start_time: string;
  end_time: string | null;
  status: string;
  duration_minutes: number | null;
  rto_achieved: boolean | null;
  rpo_achieved: boolean | null;
  validation_results: Record<string, boolean>;
  error_messages: string[];
  recommendations: string[];
  metadata: Record<string, any>;
}

export interface DRTestRequest {
  component: string;
  test_type: string;
  dry_run?: boolean;
  notify_on_completion?: boolean;
}

export interface DRTestResponse {
  test_id: string;
  status: string;
  message: string;
  estimated_duration_minutes?: number;
}

export interface DRHealthScore {
  overall_score: number;
  component_scores: Record<string, {
    score: number;
    status: string;
    priority: number;
  }>;
  status: string;
  last_updated: string;
}

export interface DREventsParams {
  limit?: number;
  severity?: string;
  component?: string;
  hours?: number;
}

export interface DRTestResultsParams {
  limit?: number;
  component?: string;
  test_type?: string;
  status?: string;
}

export interface DRFullTestParams {
  dryRun?: boolean;
  notify_on_completion?: boolean;
}

// ============================================================================
// API SERVICE CLASS
// ============================================================================

class DisasterRecoveryAPI {
  private readonly basePath = '/api/v1/disaster-recovery';

  /**
   * Get comprehensive DR status
   */
  async getStatus(): Promise<DRStatus> {
    const response = await apiClient.get(`${this.basePath}/status`);
    return response.data;
  }

  /**
   * Get detailed status of all DR components
   */
  async getComponents(): Promise<DRComponent[]> {
    const response = await apiClient.get(`${this.basePath}/components`);
    return response.data;
  }

  /**
   * Get DR events with filtering options
   */
  async getEvents(params: DREventsParams = {}): Promise<{ events: DREvent[] }> {
    const response = await apiClient.get(`${this.basePath}/events`, { params });
    return { events: response.data };
  }

  /**
   * Get DR health score and breakdown
   */
  async getHealthScore(): Promise<DRHealthScore> {
    const response = await apiClient.get(`${this.basePath}/health-score`);
    return response.data;
  }

  /**
   * Run disaster recovery test for specific component
   */
  async runTest(request: DRTestRequest): Promise<DRTestResponse> {
    const response = await apiClient.post(`${this.basePath}/test`, request);
    return response.data;
  }

  /**
   * Run comprehensive disaster recovery test for all components
   */
  async runFullTest(params: DRFullTestParams = {}): Promise<DRTestResponse> {
    const response = await apiClient.post(`${this.basePath}/test/full`, null, { params });
    return response.data;
  }

  /**
   * Get DR test results with filtering options
   */
  async getTestResults(params: DRTestResultsParams = {}): Promise<{
    test_results: DRTestResult[];
    total_count: number;
    filters_applied: Record<string, any>;
  }> {
    const response = await apiClient.get(`${this.basePath}/test/results`, { params });
    return response.data;
  }

  /**
   * Get available test types
   */
  getTestTypes(): string[] {
    return [
      'backup_validation',
      'failover_test',
      'recovery_test',
      'full_dr_test',
      'network_partition',
      'database_failover',
      'application_failover'
    ];
  }

  /**
   * Get available component types
   */
  getComponentTypes(): string[] {
    return [
      'database',
      'api_services',
      'web_ui',
      'monitoring',
      'file_storage'
    ];
  }

  /**
   * Get severity levels
   */
  getSeverityLevels(): string[] {
    return [
      'info',
      'warning',
      'critical',
      'emergency'
    ];
  }

  /**
   * Get status types
   */
  getStatusTypes(): string[] {
    return [
      'healthy',
      'warning',
      'critical',
      'testing',
      'failover_in_progress',
      'recovery_in_progress',
      'failed'
    ];
  }

  /**
   * Format duration from minutes to human readable
   */
  formatDuration(minutes: number): string {
    if (minutes < 60) {
      return `${minutes}m`;
    }
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    return remainingMinutes > 0 ? `${hours}h ${remainingMinutes}m` : `${hours}h`;
  }

  /**
   * Format relative time
   */
  formatRelativeTime(dateString: string | null): string {
    if (!dateString) return 'Never';
    
    const date = new Date(dateString);
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffMinutes = Math.floor(diffMs / (1000 * 60));
    const diffHours = Math.floor(diffMinutes / 60);
    const diffDays = Math.floor(diffHours / 24);

    if (diffDays > 0) {
      return `${diffDays}d ago`;
    } else if (diffHours > 0) {
      return `${diffHours}h ago`;
    } else if (diffMinutes > 0) {
      return `${diffMinutes}m ago`;
    } else {
      return 'Just now';
    }
  }

  /**
   * Get status color for UI components
   */
  getStatusColor(status: string): 'success' | 'warning' | 'error' | 'info' | 'default' {
    switch (status) {
      case 'healthy':
        return 'success';
      case 'warning':
        return 'warning';
      case 'critical':
      case 'failed':
        return 'error';
      case 'testing':
      case 'failover_in_progress':
      case 'recovery_in_progress':
        return 'info';
      default:
        return 'default';
    }
  }

  /**
   * Get severity color for UI components
   */
  getSeverityColor(severity: string): 'success' | 'warning' | 'error' | 'info' | 'default' {
    switch (severity) {
      case 'info':
        return 'info';
      case 'warning':
        return 'warning';
      case 'critical':
      case 'emergency':
        return 'error';
      default:
        return 'default';
    }
  }

  /**
   * Get priority label
   */
  getPriorityLabel(priority: number): string {
    switch (priority) {
      case 1:
        return 'Critical';
      case 2:
        return 'High';
      case 3:
        return 'Medium';
      case 4:
        return 'Low';
      default:
        return 'Unknown';
    }
  }

  /**
   * Get priority color
   */
  getPriorityColor(priority: number): 'error' | 'warning' | 'info' | 'default' {
    switch (priority) {
      case 1:
        return 'error';
      case 2:
        return 'warning';
      case 3:
        return 'info';
      case 4:
      default:
        return 'default';
    }
  }

  /**
   * Calculate health score color
   */
  getHealthScoreColor(score: number): 'success' | 'warning' | 'error' {
    if (score >= 80) {
      return 'success';
    } else if (score >= 60) {
      return 'warning';
    } else {
      return 'error';
    }
  }

  /**
   * Validate test request
   */
  validateTestRequest(request: DRTestRequest): string[] {
    const errors: string[] = [];

    if (!request.component) {
      errors.push('Component is required');
    }

    if (!request.test_type) {
      errors.push('Test type is required');
    }

    const validTestTypes = this.getTestTypes();
    if (request.test_type && !validTestTypes.includes(request.test_type)) {
      errors.push(`Invalid test type. Must be one of: ${validTestTypes.join(', ')}`);
    }

    return errors;
  }

  /**
   * Estimate test duration
   */
  estimateTestDuration(component: string, testType: string): number {
    const baseDurations: Record<string, number> = {
      'backup_validation': 5,
      'failover_test': 15,
      'recovery_test': 30,
      'full_dr_test': 60
    };

    const componentMultipliers: Record<string, number> = {
      'database': 1.5,
      'api_services': 1.0,
      'web_ui': 0.5,
      'monitoring': 1.2,
      'file_storage': 2.0
    };

    const baseDuration = baseDurations[testType] || 10;
    const multiplier = componentMultipliers[component] || 1.0;

    return Math.round(baseDuration * multiplier);
  }
}

// ============================================================================
// EXPORT SINGLETON INSTANCE
// ============================================================================

export const drApi = new DisasterRecoveryAPI();

File: core_infra/api/routes/disaster_recovery.py
"""
RegulensAI Disaster Recovery API Routes
Web API endpoints for DR management, testing, and monitoring integration.
"""

from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
import structlog

from core_infra.auth.dependencies import get_current_user, require_permission
from core_infra.disaster_recovery.dr_manager import (
    dr_manager,
    get_dr_status,
    run_dr_test,
    run_full_dr_test,
    DRTestType,
    DRStatus,
    DRSeverity
)
from core_infra.logging.centralized_logging import get_centralized_logger, LogCategory
from core_infra.monitoring.apm_integration import apm_manager
from core_infra.monitoring.apm_decorators import apm_track_api
from core_infra.notification.notification_service import notification_service


logger = structlog.get_logger(__name__)
router = APIRouter()


# ============================================================================
# REQUEST/RESPONSE MODELS
# ============================================================================

class DRTestRequest(BaseModel):
    """DR test execution request."""
    component: str = Field(..., description="Component to test")
    test_type: str = Field(..., description="Type of DR test")
    dry_run: bool = Field(default=True, description="Whether to run in dry-run mode")
    notify_on_completion: bool = Field(default=True, description="Send notification on completion")


class DRTestResponse(BaseModel):
    """DR test execution response."""
    test_id: str
    status: str
    message: str
    estimated_duration_minutes: Optional[int] = None


class DRStatusResponse(BaseModel):
    """DR status response."""
    overall_status: str
    health_score: float
    components: Dict[str, Any]
    recent_events: List[Dict[str, Any]]
    recent_tests: List[Dict[str, Any]]
    last_updated: str


class DREventResponse(BaseModel):
    """DR event response."""
    event_id: str
    timestamp: str
    event_type: str
    severity: str
    component: str
    description: str
    status: str
    recovery_actions: List[str]


class DRComponentResponse(BaseModel):
    """DR component status response."""
    component: str
    status: str
    rto_minutes: int
    rpo_minutes: int
    priority: int
    automated_recovery: bool
    last_test_time: Optional[str]
    last_backup_time: Optional[str]
    dependencies: List[str]


# ============================================================================
# DR STATUS AND MONITORING ENDPOINTS
# ============================================================================

@router.get("/status", response_model=DRStatusResponse)
@apm_track_api("get_dr_status")
async def get_disaster_recovery_status(
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:read"))
):
    """Get comprehensive disaster recovery status."""
    try:
        status = await get_dr_status()
        
        return DRStatusResponse(
            overall_status=status["overall_status"],
            health_score=status["health_score"],
            components=status["components"],
            recent_events=status["recent_events"],
            recent_tests=status["recent_tests"],
            last_updated=status["last_updated"]
        )
        
    except Exception as e:
        logger.error(f"Failed to get DR status: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve DR status")


@router.get("/components", response_model=List[DRComponentResponse])
@apm_track_api("get_dr_components")
async def get_dr_components(
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:read"))
):
    """Get detailed status of all DR components."""
    try:
        status = await get_dr_status()
        components = []
        
        for component_name, component_data in status["components"].items():
            components.append(DRComponentResponse(
                component=component_name,
                status=component_data["status"],
                rto_minutes=component_data["rto_minutes"],
                rpo_minutes=component_data["rpo_minutes"],
                priority=component_data["priority"],
                automated_recovery=component_data["automated_recovery"],
                last_test_time=component_data["last_test_time"],
                last_backup_time=component_data["last_backup_time"],
                dependencies=component_data["dependencies"]
            ))
        
        # Sort by priority
        components.sort(key=lambda x: x.priority)
        
        return components
        
    except Exception as e:
        logger.error(f"Failed to get DR components: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve DR components")


@router.get("/events", response_model=List[DREventResponse])
@apm_track_api("get_dr_events")
async def get_dr_events(
    limit: int = Query(default=50, le=200, description="Maximum number of events to return"),
    severity: Optional[str] = Query(default=None, description="Filter by severity"),
    component: Optional[str] = Query(default=None, description="Filter by component"),
    hours: Optional[int] = Query(default=24, description="Hours of history to include"),
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:read"))
):
    """Get DR events with filtering options."""
    try:
        status = await get_dr_status()
        events = status["recent_events"]
        
        # Apply filters
        if severity:
            events = [e for e in events if e.get("severity") == severity]
        
        if component:
            events = [e for e in events if e.get("component") == component]
        
        if hours:
            cutoff_time = datetime.utcnow() - timedelta(hours=hours)
            events = [
                e for e in events 
                if datetime.fromisoformat(e.get("timestamp", "")) > cutoff_time
            ]
        
        # Limit results
        events = events[:limit]
        
        return [
            DREventResponse(
                event_id=event["event_id"],
                timestamp=event["timestamp"],
                event_type=event["event_type"],
                severity=event["severity"],
                component=event["component"],
                description=event["description"],
                status=event["status"],
                recovery_actions=event.get("recovery_actions", [])
            )
            for event in events
        ]
        
    except Exception as e:
        logger.error(f"Failed to get DR events: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve DR events")


@router.get("/health-score")
@apm_track_api("get_dr_health_score")
async def get_dr_health_score(
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:read"))
):
    """Get DR health score and breakdown."""
    try:
        status = await get_dr_status()
        
        # Calculate component scores
        component_scores = {}
        for component_name, component_data in status["components"].items():
            if component_data["status"] == "healthy":
                score = 100
            elif component_data["status"] == "warning":
                score = 70
            elif component_data["status"] == "testing":
                score = 85
            else:
                score = 0
            
            component_scores[component_name] = {
                "score": score,
                "status": component_data["status"],
                "priority": component_data["priority"]
            }
        
        return {
            "overall_score": status["health_score"],
            "component_scores": component_scores,
            "status": status["overall_status"],
            "last_updated": status["last_updated"]
        }
        
    except Exception as e:
        logger.error(f"Failed to get DR health score: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve DR health score")


# ============================================================================
# DR TESTING ENDPOINTS
# ============================================================================

@router.post("/test", response_model=DRTestResponse)
@apm_track_api("run_dr_test")
async def run_disaster_recovery_test(
    request: DRTestRequest,
    background_tasks: BackgroundTasks,
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:test"))
):
    """Run disaster recovery test for specific component."""
    try:
        # Validate test type
        try:
            test_type = DRTestType(request.test_type)
        except ValueError:
            raise HTTPException(
                status_code=400, 
                detail=f"Invalid test type: {request.test_type}"
            )
        
        # Validate component
        status = await get_dr_status()
        if request.component not in status["components"]:
            raise HTTPException(
                status_code=400,
                detail=f"Unknown component: {request.component}"
            )
        
        # Estimate duration based on test type and component
        estimated_duration = _estimate_test_duration(request.component, test_type)
        
        # Run test in background
        background_tasks.add_task(
            _execute_dr_test_background,
            request.component,
            request.test_type,
            request.dry_run,
            request.notify_on_completion,
            current_user.id
        )
        
        # Log test initiation
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.info(
            f"DR test initiated: {request.component} - {request.test_type}",
            category=LogCategory.SYSTEM,
            component=request.component,
            test_type=request.test_type,
            dry_run=request.dry_run,
            user_id=current_user.id
        )
        
        return DRTestResponse(
            test_id=f"{request.component}_{request.test_type}_{int(datetime.utcnow().timestamp())}",
            status="initiated",
            message=f"DR test started for {request.component}",
            estimated_duration_minutes=estimated_duration
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to initiate DR test: {e}")
        raise HTTPException(status_code=500, detail="Failed to initiate DR test")


@router.post("/test/full", response_model=DRTestResponse)
@apm_track_api("run_full_dr_test")
async def run_full_disaster_recovery_test(
    dry_run: bool = Query(default=True, description="Whether to run in dry-run mode"),
    notify_on_completion: bool = Query(default=True, description="Send notification on completion"),
    background_tasks: BackgroundTasks = BackgroundTasks(),
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:test"))
):
    """Run comprehensive disaster recovery test for all components."""
    try:
        # Run full test in background
        background_tasks.add_task(
            _execute_full_dr_test_background,
            dry_run,
            notify_on_completion,
            current_user.id
        )
        
        # Log full test initiation
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.info(
            "Full DR test initiated",
            category=LogCategory.SYSTEM,
            dry_run=dry_run,
            user_id=current_user.id
        )
        
        return DRTestResponse(
            test_id=f"full_dr_test_{int(datetime.utcnow().timestamp())}",
            status="initiated",
            message="Full DR test started for all components",
            estimated_duration_minutes=60  # Estimate 1 hour for full test
        )
        
    except Exception as e:
        logger.error(f"Failed to initiate full DR test: {e}")
        raise HTTPException(status_code=500, detail="Failed to initiate full DR test")


@router.get("/test/results")
@apm_track_api("get_dr_test_results")
async def get_dr_test_results(
    limit: int = Query(default=20, le=100, description="Maximum number of results to return"),
    component: Optional[str] = Query(default=None, description="Filter by component"),
    test_type: Optional[str] = Query(default=None, description="Filter by test type"),
    status: Optional[str] = Query(default=None, description="Filter by test status"),
    current_user = Depends(get_current_user),
    _: None = Depends(require_permission("dr:read"))
):
    """Get DR test results with filtering options."""
    try:
        dr_status = await get_dr_status()
        test_results = dr_status["recent_tests"]
        
        # Apply filters
        if component:
            test_results = [t for t in test_results if t.get("component") == component]
        
        if test_type:
            test_results = [t for t in test_results if t.get("test_type") == test_type]
        
        if status:
            test_results = [t for t in test_results if t.get("status") == status]
        
        # Limit results
        test_results = test_results[:limit]
        
        return {
            "test_results": test_results,
            "total_count": len(test_results),
            "filters_applied": {
                "component": component,
                "test_type": test_type,
                "status": status
            }
        }
        
    except Exception as e:
        logger.error(f"Failed to get DR test results: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve DR test results")


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def _estimate_test_duration(component: str, test_type: DRTestType) -> int:
    """Estimate test duration in minutes."""
    base_durations = {
        DRTestType.BACKUP_VALIDATION: 5,
        DRTestType.FAILOVER_TEST: 15,
        DRTestType.RECOVERY_TEST: 30,
        DRTestType.FULL_DR_TEST: 60
    }
    
    # Adjust based on component complexity
    component_multipliers = {
        "database": 1.5,
        "api_services": 1.0,
        "web_ui": 0.5,
        "monitoring": 1.2,
        "file_storage": 2.0
    }
    
    base_duration = base_durations.get(test_type, 10)
    multiplier = component_multipliers.get(component, 1.0)
    
    return int(base_duration * multiplier)


async def _execute_dr_test_background(
    component: str,
    test_type: str,
    dry_run: bool,
    notify_on_completion: bool,
    user_id: str
):
    """Execute DR test in background task."""
    try:
        result = await run_dr_test(component, test_type, dry_run)
        
        # Log completion
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.info(
            f"DR test completed: {component} - {test_type}",
            category=LogCategory.SYSTEM,
            component=component,
            test_type=test_type,
            test_status=result.status,
            duration_minutes=result.duration_minutes,
            user_id=user_id
        )
        
        # Send notification if requested
        if notify_on_completion:
            await _send_test_completion_notification(result, user_id)
            
    except Exception as e:
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.error(
            f"DR test failed in background: {component} - {test_type}",
            category=LogCategory.SYSTEM,
            error=str(e),
            user_id=user_id
        )


async def _execute_full_dr_test_background(
    dry_run: bool,
    notify_on_completion: bool,
    user_id: str
):
    """Execute full DR test in background task."""
    try:
        results = await run_full_dr_test(dry_run)
        
        # Calculate summary
        total_tests = len(results)
        passed_tests = sum(1 for r in results.values() if r.status == "passed")
        failed_tests = total_tests - passed_tests
        
        # Log completion
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.info(
            "Full DR test completed",
            category=LogCategory.SYSTEM,
            total_tests=total_tests,
            passed_tests=passed_tests,
            failed_tests=failed_tests,
            user_id=user_id
        )
        
        # Send notification if requested
        if notify_on_completion:
            await _send_full_test_completion_notification(results, user_id)
            
    except Exception as e:
        dr_logger = await get_centralized_logger("dr_api")
        await dr_logger.error(
            "Full DR test failed in background",
            category=LogCategory.SYSTEM,
            error=str(e),
            user_id=user_id
        )


async def _send_test_completion_notification(result, user_id: str):
    """Send notification for test completion."""
    await notification_service.send('DR Test Completed', f'Status: {result.status}', user_id)


async def _send_full_test_completion_notification(results: Dict, user_id: str):
    """Send notification for full test completion."""
    summary = f"Full DR test completed with {len(results)} components tested."
    await notification_service.send('Full DR Test Completed', summary, user_id)
